### ISSUE TITLE
 Internal Server Error - https://<app_name>.azurewebsites.net/admin/host/keys

### ISSUE THREAD

**Issue by btardif** [Azure/Azure-Functions#719]
_From @handeeadiguzel on February 12, 2018 15:1_

> Please provide us with the following information:
> ---------------------------------------------------------------

### OS and Browser
> Windows & Chrome 

### Repro steps
> 
1. Create function app from azure portal
2. Change runtime version to beta (2.0.11415.0 (beta))
2. Enable deployment slots

Host keys are gone (portal doesn't show them anymore). And cannot add new keys. 

### Any error messages
>

> Internal Server Error - https://<app_name>/admin/host/keys
> Session Id: 40a3ea0af19f42ceae950c973d66b770
> 
> Timestamp: 2018-02-12T14:42:19.631Z

### Mention any other details that might be useful.

_Copied from original issue: Azure/azure-functions-ux#2334_

**Comment by btardif** [comment]
@ahmelsayed can you take a look?

**Comment by Neonkiller** [comment]
Hello
I'm struggling the same issue. But runtime version is 1. It just stopped working. Any updates on it?

**Comment by APErebus** [comment]
I'd like to get some more info on this as well. We are getting this after creating a Azure Function via Terraform & deploying via Octopus Deploy.

This was working perfectly fine a week ago, now it's just failing continually

**Comment by worldspawn** [comment]
I'm now getting this as well on a function after deploying a new release. :'(

![image](https://user-images.githubusercontent.com/819581/41915325-c9c2aae0-7998-11e8-8964-06b90a81c0f3.png)

```
Error:

Internal Server Error - https://<name>.azurewebsites.net/admin/host/keys
Session Id: e2207e819afe493687b65cf4b1441f71

Timestamp: 2018-06-26T13:21:21.196Z
```

```
Error:

Internal Server Error - https://<name>.azurewebsites.net/admin/host/systemkeys/_master
Session Id: e2207e819afe493687b65cf4b1441f71

Timestamp: 2018-06-26T13:23:12.606Z
```

**Comment by btardif** [comment]
@paulbatum  can you triage this item?

**Comment by worldspawn** [comment]
@btardif We have identified a possible cause for this. We were creating our function app using the terraform azure provider which seems to also create a few basic settings. Two of those settings are:
`WEBSITE_CONTENTSHARE`
`WEBSITE_CONTENTAZUREFILECONNECTIONSTRING`

These settings arn't part of our deployment so they are being removed when we deploy. The act of removing them seems to break the function app. I'm told these settings are intended for consumption plan backed functions only.

We couldn't hack terraform into not doing that so we've just added them to our set of dev settings until the provider can be patched.

**Comment by Gowiem** [comment]
Any final resolution on this issue? I'm running into this when deploying a fresh Function App with a `FUNCTIONS_EXTENSION_VERSION: beta` AppSetting.

**Comment by btardif** [comment]
@ahmelsayed @paulbatum can you guys comment on this?

**Comment by Davy-F** [comment]
I'm also getting '500 Internal Server Error' when deploying a new, v2 C# Function App.
I've even tried doing a simple Hello World app and still get the same behaviour.

Any updates on this?

**Comment by SteveVoros** [comment]
I had the same problem. It was caused because I applied additional App Settings configurations in a second step using PowerShell where I accidentally wiped out default App Settings. After that, redeployment of the Function App failed and the only fix was to delete and recreate the Function App.

**Comment by Gowiem** [comment]
@btardif Continuing to run into this issue with a function app deployed with `FUNCTIONS_EXTENSION_VERSION: 2.0.11961-alpha`.

**Comment by kazinad** [comment]
If you delete your function app+plan, don't forget to delete the storage account file share as well. That fixed my problem

**Comment by pragathikadaganchi** [comment]
I'm facing the same issue after being delpoyed new function app to azure with java runtime

Error:

Internal Server Error - https://scanandgo-warmup.azurewebsites.net/admin/host/keys
Session Id: ae893841c0764d98bc44b939653295c5

Timestamp: 2018-09-09T03:03:00.461Z

**Comment by xtellurian** [comment]
Facing the same issue.

Steps to repro:
1. Create new Functions App on Windows with App Service Plan.
2. Change runtime to V2 
3. Create a new JS HTTP trigger function.

Expected result: Things are OK.

Actual Result: Host Keys are deleted and can't be created.

```
Error:

Internal Server Error - https://locafi-ingress-2.azurewebsites.net/admin/host/systemkeys/_master

Session Id: 9cb3c262b262401096ad8aaebf637e03

Timestamp: 2018-09-12T05:02:45.815Z
```

**Comment by jeromekol** [comment]
We are facing the same issue as @xtellurian

**Comment by jeromekol** [comment]
We have the same issue but with c# http trigger functions

**Comment by paulbatum** [comment]
I am not able to reproduce, in general. I investigated a few cases where the function app name and timestamp were shared.

@xtellurian Your function app is hitting outbound connection exhaustion errors. It looks like the connections are being exhausted by other applications running on your same app service plan. I am not able to tell which application is doing it, but I can see there are a huge amount of connections (~1900) to a specific IP address and port. If you need additional details and you're OK with me sharing the full IP and port, let me know and I can do so.

@pragathikadaganchi I'm seeing a similar case to the one above - errors relating to outbound connection exhaustion. In your case I see ~500 connections to a specific IP and port. Again, if you're OK with me sharing full details, let me know and I will do so.

I have filed https://github.com/Azure/azure-functions-host/issues/3444 to track some improvements that can be made.

Also, it looks like there are additional logs that are missing in version 12050 that would be helpful in narrowing down further why exactly these connection issues are impacting the master key API. We have fixed the issue that is preventing these logs from being present. So once the new version of functions is deployed, we can take another look at the logs for apps that are still hitting this error, and they should help further troubleshooting.

**Comment by xtellurian** [comment]
@paulbatum Ah, OK that's a bug, I believe it's from not re-using the connection to IoT Hub (the function is acting as a proxy for HTTP messages into IoT Hub). 

I'll try to repro in a new App Service

**Comment by chris-rogala** [comment]
Any update on this?  I just ran into the same error today.

**Comment by paulbatum** [comment]
This investigation has stalled. Cases I looked at were due to other problems (see my recent reply above). If you post detailed information about your particular case we can take a look at our logs.

**Comment by ScottKriegerPD** [comment]
I'm running into this with a new Azure function app. 

Internal Server Error - https://api-inscape.azurewebsites.net/admin/host/keys
Session Id: 36e8a8ca2d4146eebabe5dc3aaad58d7

Timestamp: 2018-11-09T15:38:25.898Z

**Comment by ScottKriegerPD** [comment]
Thanks @paulbatum. I recreated the app and it's functioning now. I'll watch for that in the future.

**Comment by sprabhus** [comment]
I am too running into same issue with new the azure function app where I tried all the observations posted above.

Error:
Internal Server Error - https://oneprofilesyncappdev.azurewebsites.net/admin/host/systemkeys/_master
Session Id: a76915240e404f67b58729ae68be9c6a

Timestamp: 2018-11-28T09:12:45.049Z

**Comment by paulbatum** [comment]
@sprabhus I looked at your case, I see an error I have not seen before:

```
Microsoft.WindowsAzure.Storage.StorageException : One of the request inputs is out of range.
   at async Microsoft.WindowsAzure.Storage.Core.Executor.Executor.ExecuteAsyncInternal[T](RESTCommand`1 cmd,IRetryPolicy policy,OperationContext operationContext,CancellationToken token)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.WindowsAzure.Storage.Blob.CloudBlobContainer.CreateIfNotExistsAsync(??)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.Azure.WebJobs.Script.WebHost.BlobStorageSecretsRepository..ctor(String secretSentinelDirectoryPath,String accountConnectionString,String siteSlotName) at C:\azure-webjobs-sdk-script\src\WebJobs.Script.WebHost\Security\KeyManagement\BlobStorageSecretsRepository.cs : 65
   at Microsoft.Azure.WebJobs.Script.WebHost.DefaultSecretManagerProvider.CreateSecretsRepository() at C:\azure-webjobs-sdk-script\src\WebJobs.Script.WebHost\Security\KeyManagement\DefaultSecretManagerProvider.cs : 66
   at Microsoft.Azure.WebJobs.Script.WebHost.DefaultSecretManagerProvider.Create() at C:\azure-webjobs-sdk-script\src\WebJobs.Script.WebHost\Security\KeyManagement\DefaultSecretManagerProvider.cs : 48
```

@alrod any ideas about what might cause this?

**Comment by alrod** [comment]
@sprabhus can you please check "AzureWebJobsStorage" app setting. Can you connect to the storage using "Azure Storage Explorer"?

**Comment by sprabhus** [comment]
Yes AzureWebJobsStorage has the  connection string to the storage account name-opSyncAppStorageDEV and i am able to connect to the storage account using storage explorer using the connection string in the AzureWebJobsStorage app setting. Please let me know if you need any further details.

**Comment by MysticWolf** [comment]
Hello,
I got the same issue with an azure function created from Visual Code. My function was using a trigger from azure services bus. Integration part was badly created by the function. The connection string was missing related to the queue and in the "Integrate" tab the connection was empty (but not topic and subscription...). After, adding again the creation again the connection from "integrate" it worked.

**Comment by rtaylor72** [comment]
I am also having this error and I have "AzureWebJobsStorage" specified.
```
Error:
Internal Server Error - https://integrationhubcorefunction.azurewebsites.net/admin/host/systemkeys/_master 
Session Id: 53df7b45cc2f4a83a50449316e46a9ec
Timestamp: 2018-12-10T17:15:29.650Z
```

**Comment by alrod** [comment]
@rtaylor72 
I saw storage related errors in logs around 2018-12-1017:30:00 GMT
*System.InvalidOperationException : Secret initialization from Blob storage failed due to a missing Azure Storage connection string. If you intend to use files for secrets, add an App Setting key * 'AzureWebJobsSecretStorageType' with value 'Files'
Microsoft.WindowsAzure.Storage.StorageException : This request is not authorized to perform this operation.
I looks like you fixed this issue. Latest error was in function code itself:
Microsoft.Azure.WebJobs.Host.FunctionInvocationException : Exception while executing function: CreateTransaction ---> System.ArgumentNullException : Value cannot be null.

**Comment by alrod** [comment]
@sprabhus 
I saw an error in the logs:
System.ArgumentException : Value for the connection string parameter name [Guid] was not found.
Parameter name: connectionString

I looks like you fixed this issue. Latest error is Timeout value of 00:30:00 was exceeded by function: Re..nc
Can you confirm that?

**Comment by sprabhus** [comment]
Yes, you are right. I figure out that connection string for one of the function was missing and updated that last week.

Inspite of that I am still receiving the same error,

**Error:

Internal Server Error - https://oneprofilesyncappdev.azurewebsites.net/admin/host/systemkeys/_master
Session Id: 79ae11f203a444898151cd3a88ecf319

Timestamp: 2018-12-12T09:42:50.502Z**

**Comment by alrod** [comment]
@sprabhus, can you please open "Blob Containers" node in Azure Storage Explorer in your storage account. Do you see any containers?

**Comment by sprabhus** [comment]
@alrod  Yes, I am able to connect to the storage account and i don't see any blob container in that account. Please find the screen shot for the same.

![storagecontainer](https://user-images.githubusercontent.com/13001718/50343014-71b6c400-054b-11e9-8359-21f000259c8f.png)

**Comment by TechnologyAnimal** [comment]
Same problem for me...  I, too, have "AzureWebJobsStorage" defined.

Error:

Internal Server Error - https://prod-des-healthcheck.azurewebsites.net/admin/host/keys
Session Id: 70dc211828e2463197daf4b5f70a74a4

Timestamp: 2018-12-30T05:43:34.760Z

**Comment by thomas-schreiter** [comment]
Same problem here. I use Azure Functions v2 with slot swapping (i.e. deploy to staging slot then swap the slots.)

_Internal Server Error - https://expsummarysandbox-func-dev-wus.azurewebsites.net/admin/host/keys
Session Id: 486410cb3fa54f1d969e6a86234833f6_

_Timestamp: 2019-01-04T19:23:01.216Z_

More specifically, I just upgraded the Functions version from v1 to v2 when this issue started occurring.  What's funny is that the error occurs only on the production slot on alternating deploys -- i.e. on deploy _n_ production is broken and staging works, on deploy _n+1_ both production and staging work, on deploy _n+2_ production is broken again and so forth.

Is there a way to hard-reset the keys, e.g. by deleting a blob or something like that? If nothing else works, I will have to recreate the whole function app, but I rather avoid that because I have a couple of environments where I will have to repeat this process.

I tried deploying to the production slot directly, but the error still occurs. 

Any ideas, @paulbatum ?

**Comment by paulbatum** [comment]
@thomas-schreiter this information is really useful, thank you. Just to confirm, in your case does each deployment consist of deploying to staging and then doing a swap?

Yes you can hard delete keys and force them to be regenerated. By default they are in a blob container called 'azure-webjobs-secrets' in the storage account used by your function app. If you are sharing that storage account across multiple function apps, be careful as there will be multiple folders within that container. The host.json file within the folder has host level keys, and then there are json files per function.

**Comment by thomas-schreiter** [comment]
Thanks for the quick response, @paulbatum. I deleted the host.json. It got recreated after startup, but the errors still occurs. 

About your question: Our build pipeline first deploys to staging and then swaps the slot. Now, for debugging this issue, I tried deploying to production directly (i.e. without slot swapping), but it didn't resolve the issue.

**Comment by thomas-schreiter** [comment]
@paulbatum : I looked further, but didn't find too much. Maybe this bit helps you to find the root cause: I store secrets in a Key Vault and set app settings with a string like "@Microsoft.KeyVault(SecretUri=https://secret_uri_with_version". In Kudu Environment, these strings don't get resolved in the broken production slot. In the staging slot, they do get resolved. Furthermore, the production slot tries to start up but quickly fails when trying to connect to the Azure Service Bus, most likely because the secret wasn't retrieved from the Key Vault.

**Comment by paulbatum** [comment]
I have set up a repro environment, but I am getting some other type of error when I do a swap (not the same as what you're getting). I'm trying to figure out if I've screwed something up myself or if its a different bug. Probably won't get to the bottom of this until next week.

**Comment by thomas-schreiter** [comment]
I tried another thing without success: Deploy the old Azure Functions v1 again and then deploy v2. Same error. I noticed after deploying v1, that the secrets still don't get resolved. @paulbatum, let me know if you want to look further into this today. Otherwise I'm going to delete the app and start from scratch to unblock myself.

**Comment by paulbatum** [comment]
Go ahead and unblock yourself. I have the logs from the errors that your app was encountering and I'm still trying to get a repro.

**Comment by thomas-schreiter** [comment]
I deleted the app and the underlying storage account and then recreated them both. At first, the error occurred again. Then I realized that the Key Vault needs to authorize to the new app, because it got a new app id (same name, but different id). Everything works now. Looks like everything is working now.

I have a feature request: If the secrets in the app settings can't be resolved, then return a error message saying that, instead of the error message of OP. In my case the function app wasn't authorized to read secrets from the Key Vault.

**Comment by thomas-schreiter** [comment]
I looked a little bit more. I believe it originally broke when I tried to swap slots that were on different versions. Production was on v1, Staging was on v2, and swapping them somehow messed things up. What does seem to work is first updating both slots individually to v2 -- then it's possible to swap again.

**Comment by paulbatum** [comment]
Ahh, thats interesting. I assume when you set this up, you made sure that FUNCTIONS_EXTENSION_VERSION wasnt a sticky setting, and therefore its value swapped as part of the swap operation? Because as long as that is done correctly, I feel like this scenario should work (and perhaps there is a bug that we need to find and fix to make it so that it does work).

**Comment by thomas-schreiter** [comment]
That's right. The app settings including FUNCTIONS_EXTENSION_VERSION are set in the CD pipeline via a Powershell script. All except for one setting are non-sticky -- and the sticky one is not related to the Functions framework.

**Comment by paulbatum** [comment]
So I figured out what was wrong with my attempt to repro. I was hitting a different issue due to a config mistake I made. I am able to successfully do swaps between two slots for a function app that is using app setting references to key vault and is using MSI to authenticate with the key vault, without any errors.

This backs up the theory we were discussing above, that its related to mixed versions and is not a MSI + Key Vault + Slots issue.

**Comment by dantheman999301** [comment]
I've got this problem, but I do not have any slots or keys.

App name is ukwest-bill-proc-7e56185c9cd1.

Session id is 4232786c13c6463d982c35f1032b5423

It's a durable function any also have a Service Bus Trigger like the one above.

I've tried completely recreating all resources from scratch (storage, plan and function app) thinking something had got messed up along the way but no dice there.

Happy to provide any more information.

**Comment by ivana61** [comment]
I am getting this error: 
Error:
Not Found - https://xxxx.azurewebsites.net/admin/host/keys

I am creating the resource through an ARM template with a release pipeline. The host keys cannot be found. Any idea how can I solve this?
I can provide more info if needed.

**Comment by MarcelMeurer** [comment]
Hi.

Take a look here: https://github.com/Azure/Azure-Functions/issues/1118

Marcel

Outlook für Android<https://aka.ms/ghei36> herunterladen

________________________________

**Comment by ivana61** [comment]
Didn't help, I miss the host keys when it is deployed with ARM through Azure DevOps release pipeline with the task Azure Deployment: Create or Update Resource Group.

![image](https://user-images.githubusercontent.com/37874763/53327799-d514f780-38e8-11e9-8370-5dc1d7ee1fa4.png)

Any suggestions where it goes wrong?

**Comment by audipen** [comment]
I get a similar error - 
**Internal Server Error - https://btcamhsettingsfunceudly.azurewebsites.net/admin/functions/SettingGet/keys**

Since my functions are HTTP triggered and I want to have only the required app settings, I do not have the 'AzureWebJobsStorage' app setting. [This link](https://docs.microsoft.com/en-us/azure/azure-functions/functions-app-settings#azurewebjobsstorage) specifically mentions '**_The Azure Functions runtime uses this storage account connection string for all functions except for HTTP triggered functions._**'

Is this no longer true for v2 functions? Should I always have this app setting?

**Comment by alrod** [comment]
@audipen 
**Keys for calling HTTP endpoints are always stored encrypted in Azure Blob storage.** In version 1.x, keys were stored in Azure File storage be default. When upgrading an app from version 1.x to version 2.x, existing secrets that are in file storage are reset.
https://docs.microsoft.com/en-us/azure/azure-functions/functions-versions#changes-in-features-and-functionality

So you need 'AzureWebJobsStorage' app setting.

**Comment by alrod** [comment]
@ivana61, can you please check if a storage pointed in 'AzureWebJobsStorage' exists

**Comment by ivana61** [comment]
@alrod, yes it exists but it's empty. And when I create a function app through the portal the host keys are also created and stored in the blob storage.

**Comment by alrod** [comment]
@ivana61, what do you see when you go to app root ? 
https://yourappname.azurewebsites.net

can you share your function app name privatlly?
https://github.com/Azure/azure-functions-host/wiki/Sharing-Your-Function-App-name-privately

**Comment by ivana61** [comment]
@alrod, when I go to the app root I get "Forbidden You don't have permission to access / on this server."

Session Id: 4cb5e284d6304ea78863fcb6885e8950
Timestamp: 2019-02-25T23:04:02.300Z
West Europe

**Comment by alrod** [comment]
@ivana61, is "Session Id: 4cb5e284d6304ea78863fcb6885e8950" from "portal red bubble"? 

If you can't get the id from the function logs you can share the app privately using:
https://github.com/projectkudu/kudu/wiki/Reporting-your-site-name-without-posting-it-publicly
Both "Sharing partial information" or "Sharing via a dummy site" work

**Comment by ivana61** [comment]
@alrod, Yes that was from the red error bubble. here is my partial information, subscription: 0494........d60a3, function name: nmb.....tion and there is a text file in \home\logFiles\temp.txt.

**Comment by ivana61** [comment]
@alrod, any updates on how this can be fixed?

**Comment by alrod** [comment]
@ivana61, 
I found:
Sub:0494........d60a3
Name:nmb.....**app**
Region:West Europe
linux + python

But there is no `\home\logFiles\temp.txt`
Is it the app with the keys error?

**Comment by Izzmo** [comment]
We are also running into this issue with an App Service Plan, on Windows. You can use the function just fine, but the authentication piece fails where it tries to validate a key. Further, Azure Portal is unable to load/view the keys with an error, as described.

We were able to find a way around this by creating a separate storage account and redirecting the azure function. Pretty infuriating issue to debug.

**Comment by nickstaggs** [comment]
Anyone still having this issue I would suggest creating a new storage account for the function as @Izzmo pointed out. My organization had a number of functions sitting on the same storage account which may have contributed to the issue. Is there any guidance from Microsoft on whether multiple, different functions can sit on the same storage account @alrod ?

**Comment by paulbatum** [comment]
Our general guidance is that yes you can share storage accounts across function apps but we recommend for production deployments that you use a separate account per app.  We'll take a look at any bugs that impact the user experience when sharing a storage account. I'm not sure if any of the details in this issue point to a specific underlying bug though. If anyone has reliable repro steps that demonstrate an error when sharing storage accounts, please share them!

**Comment by jaffadan** [comment]
Is this guidance documented anywhere.  I have been given the exact opposite advice in the past since there are hard limits to the number of storage accounts per subscription.

**Comment by paulbatum** [comment]
Well if you have a scenario where you are creating thousands of function apps in a given subscription you might indeed need to share storage accounts to address limits of how many storage accounts you can create per sub. However most customers do not fall into this bucket, and so our advice is to keep each app isolated by not sharing storage accounts.

I have searched a bit in the docs but could not find this stated anywhere explicitly. Our creation flow does naturally lead users this way (when you make a new function app in the portal, the default is to make a new storage account).

**Comment by nickstaggs** [comment]
@paulbatum I don't think you will be able to find repro steps as it seems fairly intermittent. I can confirm that if you either put the problematic function app in a new storage account or delete the files on the storage account pertaining to the function app and redeploy it the error message goes away. Hopefully, this helps anyone else who happens to run across this.

**Comment by jaffadan** [comment]
@paulbatum   Could you get this sort of "Advice" documented.   These are the sort of things that would be very helpful to have in a single place.   You can add to it never use Blob triggers because they are not guaranteed to fire if the function app is not running, where an Storage Event is.

**Comment by paulbatum** [comment]
I filed a documentation issue regarding storage guidelines:
https://github.com/Azure/Azure-Functions/issues/1180

**Comment by nickstaggs** [comment]
Thanks @paulbatum!

**Comment by christianreddington** [comment]
I seem to be hitting the same issue - Similar to others, My Function App infra is deployed as an ARM template and I deploy the ARM Template using Azure DevOps, with a separate task to deploy the code/config to the Function App. I've tried a separate storage account, which also yields the same result. I can confirm I have the AzureWebJobsStorage, and have tested that the storage connection is valid in Azure Storage explorer. I can't see any blob containers, but can see the folder in the File Share.

Internal Server Error - https://th-show-dev-func.azurewebsites.net/admin/host/systemkeys/_master
Session Id: f622f8cad51b417eaf9502248ab52f6e

Timestamp: 2019-05-05T16:19:07.401Z

**Comment by christianreddington** [comment]
> I seem to be hitting the same issue - Similar to others, My Function App infra is deployed as an ARM template and I deploy the ARM Template using Azure DevOps, with a separate task to deploy the code/config to the Function App. I've tried a separate storage account, which also yields the same result. I can confirm I have the AzureWebJobsStorage, and have tested that the storage connection is valid in Azure Storage explorer. I can't see any blob containers, but can see the folder in the File Share.
> 
> Internal Server Error - https://th-show-dev-func.azurewebsites.net/admin/host/systemkeys/_master
> Session Id: f622f8cad51b417eaf9502248ab52f6e
> 
> Timestamp: 2019-05-05T16:19:07.401Z

Found my issue. I use a connection string for a Service Bus Topic, and directly pass the connection string from my ARM template into KeyVault. My Azure DevOps pipeline then passes the connection string in the deployment task. The challenge is that the ARM template also includes the "entityPath" segment at the end of the connection string. The function host complains about this if you set it locally, but not if you deploy it to Azure Functions directly. Instead, i got the error listed previously. I'll need to adjust my ARM template to cut off the piece at the end, to continue the end to end automated deployment approach. Hope this helps someone else!

**Comment by paulbatum** [comment]
@christianreddington thanks for letting us know. BTW the issue you ran into is tracked [here](https://github.com/Azure/azure-webjobs-sdk/issues/1168) and I'm trying to get a fix for it prioritized.

**Comment by christianreddington** [comment]
Awesome, thanks for the update @paulbatum

**Comment by bhazen** [comment]
I'm running into this issue as well, though it seems to be caused by the inclusion of an authorization.json file in our function app. We've included that as we'd like to have a health check endpoint anonymous and require authentication on all our other endpoints. We took the authorization.json approach based on the answer in this SO post: https://stackoverflow.com/questions/48712457/is-it-possible-to-disable-aad-auth-on-specific-azure-functions. Prior to including that file we had no issues with host keys.

The red bubble info:
Session Id: b74e942466f34ed586122ae41ee4025e
Timestamp: 2019-05-10T15:08:18.141Z

Other info for identifying function app:
2019-05-10T15:20:08.545 [Information] Executing 'Ping' (Reason='This function was programmatically called via the host APIs.', Id=3a017320-d455-4548-a7a4-f789d0552edc)
2019-05-10T15:20:08.610 [Information] Executed 'Ping' (Succeeded, Id=3a017320-d455-4548-a7a4-f789d0552edc)
Region: East US 2

**Comment by sladeedmonds** [comment]
I just ran in to this same issue when re-deploying a function to an ASE.  It was previously running extension ~1 and I needed to switch it to ~2.  I completely removed the app and re-deployed it, only adding FUNCTIONS_EXTENSION_VERSION ~2 as an appSettings in my ARM template.  The keys were missing, as others have noted.  

I then added the appSettings AzureWebJobsStorage and referenced a storage key.  So essentially, I added this to the ARM template (sorry for the poor formatting):

`
"siteConfig": {
                  "appSettings":[
                      {
                          "name": "FUNCTIONS_EXTENSION_VERSION",
                          "value": "~2"
                      },
                      {
                          "name": "AzureWebJobsStorage",
                          "value": "[concat('DefaultEndpointsProtocol=https;AccountName=', parameters('storageAccountName'), ';AccountKey=', listKeys(variables('storageAccountid'),'2015-05-01-preview').key1)]",
                      }
                  ]
              }
`

**Comment by worldspawn** [comment]
Here i formatted it for buddy 😄 👍 

``` js
{
   "siteConfig":{
      "appSettings":[
         {
            "name":"FUNCTIONS_EXTENSION_VERSION",
            "value":"~2"
         },
         {
            "name":"AzureWebJobsStorage",
            "value":"[concat('DefaultEndpointsProtocol=https;AccountName=', parameters('storageAccountName'), ';AccountKey=', listKeys(variables('storageAccountid'),'2015-05-01-preview').key1)]"
         }
      ]
   }
}
```

**Comment by joshbouganim** [comment]
Hello,

I went through this whole thread and I still don't understand how to get more information ourselves on the errors and why the host keys get deleted.  Anyway... here is my situtation, just a vanilla functions app deployed by azure dev ops

```
Error:

The function runtime is unable to start.
Session Id: d255cdb380b941d9920cc7d7add7a124

Timestamp: 2019-05-24T13:30:14.374Z

Error:

Internal Server Error - https://xxx.azurewebsites.net/admin/host/systemkeys/_master
Session Id: d255cdb380b941d9920cc7d7add7a124

Timestamp: 2019-05-24T13:30:33.152Z
```

**Comment by AlexPiksel** [comment]
I was hitting this problem with Functions accounts deployed via ARM, but not via the Portal.

I eventually fixed it by bumping the `apiVersion` for all components in my template up to the latest from [the Template resource listings](https://docs.microsoft.com/en-us/azure/templates/) . My ARM template was based on content from the Azure Quickstart GitHub, but some of the `apiVersion` values they use are very out of date. 

I'm guessing the portal automatically uses the latest APIs, which would explain why it wasn't hitting the problem when functions were created via that route.

**Comment by garg7pec** [comment]
In my function app, I created a new slot ("canary") and copied the app settings from the "production" slot.
Now, when I am trying to hit the REST API of "Canary" from Postman, it's giving me "500 Internal Server Error" whereas when I am trying to hit REST API of "Production", it's going perfectly fine.
I checked the access code and URL for Canary and they seem to be correct.
Can someone please help here?

**Comment by paulbatum** [comment]
@garg7pec The error logs should give you a clue. https://docs.microsoft.com/en-us/azure/azure-functions/functions-monitoring

**Comment by garg7pec** [comment]
@paulbatum Thanks Paul for taking out time for this. I actually checked the logs and it was only telling me that "failure : Host.Executor". I was not able to derive any meaningful info from this.
I actually added some more logs and returned the exception to Postman from where I could be able to find the issue.
Issue was with MSI.
I got the resolution as there was an issue with MSI of the slot where I need to separately create MSI for slots and add it in the KeyVault access policies.

**Comment by johnny-chan** [comment]
I've read through this thread and i'm getting confused by all the different issues that may be the cause of my issue. 

I'm trying to create a function key for my http triggered Azure function. I can't create the key from the portal and  when i've tried duing deployment  process, i get the following error

```
The remote server returned an error: (500) Internal Server Error.
``` 

I have a Powershell script with a section that invokes the key management api . 

$keys = Invoke-RestMethod -Method PUT -Headers @{Authorization=("Bearer {0}" -f $accessToken )} -ContentType "application/json" -Uri "$siteBaseUrl/admin/functions/$functionName/keys/$funcKeyName" -body $json


The error i get when viewing the azure function "manage" section is 
```
Error:

Internal Server Error - https://artio-dev-claims-service.azurewebsites.net/admin/host/systemkeys/_master
Session Id: 300fcae68d834d0981d9aee6179ec4cd

Timestamp: 2019-06-24T15:07:20.770Z
```

Any ideas would be really appreciative.

**Comment by Naetmul** [comment]
I created an Azure function (HTTP Trigger) in VS code TypeScript, following the Quickstarts page of the Azure docs, and that error came out.

My case was due to the `AzureWebJobsStorage` setting.
The default value for `AzureWebJobsStorage` in `local.settings.json` was an empty string. This can be updated into the application setting in the server when deployed.

What I did was, set the `AzureWebJobsStorage` to somewhat default value.
Specifically, in the Azure portal, I clicked the "Storage accounts" in the left panel, selected the storage account item for the function app, and clicked the "Access keys". There are strings called "Connection string" whose value is `DefaultEndPointsProtocol=https,AccountName=...`.
I copied one of the value and then pasted it into the `AzureWebJobsStorage` in `local.settings.json`, and then I deployed the function with setting update. The errors are gone now.

**Comment by johnny-chan** [comment]
Thank you so much @Naetmul! i added the AzureWebJobsStorage  setting and the error has now gone and i am able to create function keys for my azure function

**Comment by leefyone** [comment]
I have a similar issue. Our function was not working. It turned out that we too had no keys set, and the error on the Manage tab, for the relevant function was:

> Error:
> 
> Internal Server Error - https://<name>azurewebsites.net/admin/host/systemkeys/_master
> Session Id: 0cab5ad1725b4f35b6521e156fcef051
> 
> Timestamp: 2019-07-11T10:45:24.724Z

We could not set any keys, If we did we got the above error again.

Initially we fixed this by redeploying the functions again (via Octopus). Then when we looked at the keys, the Host Keys were populated. We then created a new Function Key and we use that as the API key in the function URL. When the function was called via the URL (with the api/function key) everything worked.

However, today the functions are not working again. Looking at the Manage tab again (in Function App in the portal) we have the same Internal Server Error, and all the keys are missing.

I believe all our of settings are correct (having read through this page) - also the functions worked correctly after redeployment (but only for a short time).

How do I find any logs that might indicate the issue we are having.

And why would it work initially only to stop later on.

Many thanks

**Comment by Izzmo** [comment]
We are still running into this on our side -- but I think I found a hacky fix.

## Situation / Environment
- You have an isolated ASE
- You have an ASP
- You have an azure function attached to the ASP
- You have a storage account, created with the function
- The storage account has firewall rules and peer via ExpressRoute VPN

## The problem
The problem is, you cannot access the host keys because the Azure portal cannot connect to the storage account and the azure function cannot connect to the storage account.

## Solution
1. You need to disable firewall rules on the azure function storage account where the keys are
2. You need to restart the azure function
3. Re-enable the firewall rules on the azure function storage account.

It's super weird and I don't like it.. but it works. This seems to be a super specific complicated issue very few will have since you have to be using an isolated ASE with an express route and peer networks.

**Comment by djsowa** [comment]
It seems that when having storage account with firewall and enabled option "Allow trusted Microsoft services to access this storage account", Azure Function will throw runtime error even when function and storage account are in same region.

So Azure Functions are not trusted Microsoft service ;)

**Comment by sladeedmonds** [comment]
> It seems that when having storage account with firewall and enabled option "Allow trusted Microsoft services to access this storage account", Azure Function will throw runtime error even when function and storage account are in same region.
> 
> So Azure Functions are not trusted Microsoft service ;)

Correct, the list of trusted services can be found at [https://docs.microsoft.com/en-us/azure/storage/common/storage-network-security#trusted-microsoft-services](url).

**Comment by paulbatum** [comment]
Yup. You'll notice that no compute services are in that list i.e. Azure VMs, Cloud Services, ACI, AKS, App Service, Service Fabric, Functions, etc are all missing. That's because these services will execute code specified by a user.

**Comment by redshiftz** [comment]
This also worked for me when we ran into this issue. 

> We are still running into this on our side -- but I think I found a hacky fix.
> 
> ## Situation / Environment
> * You have an isolated ASE
> * You have an ASP
> * You have an azure function attached to the ASP
> * You have a storage account, created with the function
> * The storage account has firewall rules and peer via ExpressRoute VPN
> 
> ## The problem
> The problem is, you cannot access the host keys because the Azure portal cannot connect to the storage account and the azure function cannot connect to the storage account.
> 
> ## Solution
> 1. You need to disable firewall rules on the azure function storage account where the keys are
> 2. You need to restart the azure function
> 3. Re-enable the firewall rules on the azure function storage account.
> 
> It's super weird and I don't like it.. but it works. This seems to be a super specific complicated issue very few will have since you have to be using an isolated ASE with an express route and peer networks.

**Comment by RanjithEswaran-Serverless360** [comment]
I tried deploying an Azure Function App using ARM template in Azure US Government.
I am not able to list or create the keys in the Function App. I am getting the **internal server error** whenever I try to do the above operations.  

I am unable to proceed further because of this. I have tried all the above suggestions. When i looked into the storage account configured with this function app, I couldn't see any containers created for this function app in the storage accounts. But I don't specify any network restrictions in the storage account, while deploying the storage account and Azure Function App.

Any insights on this would be of great help.

**Comment by djsowa** [comment]
> Yup. You'll notice that no compute services are in that list i.e. Azure VMs, Cloud Services, ACI, AKS, App Service, Service Fabric, Functions, etc are all missing. That's because these services will execute code specified by a user.

I can't agree, Functions needs this storage account for it's own runtime, not to interact with user code. So for this purpose Functions should be trusted when connecting to SA.
